commit 113ea1a0cdc0178a218adc1a5889b1ac54f28c7c
Author: sveinung <sveinung@a0f10bec-cc02-0410-94fc-a9cfff90b4cd>
Date:   Thu Dec 29 16:28:29 2016 +0000

    Stop checking if "any action" has a target.
    
    Only a specific action can be impossible because no suitable target is
    found.
    
    Reported by wotan <wotan>
    
    See bug #25401
    
    git-svn-id: svn://svn.gna.org/svn/freeciv/trunk@34748 a0f10bec-cc02-0410-94fc-a9cfff90b4cd

diff --git a/server/unithand.c b/server/unithand.c
index e57f20f..8b7a282 100644
--- a/server/unithand.c
+++ b/server/unithand.c
@@ -802,29 +802,32 @@ static struct ane_expl *expl_act_not_enabl(struct unit *punit,
   bool on_native = is_native_tile(unit_type_get(punit), unit_tile(punit));
   int action_custom;
 
-  switch (action_id_get_target_kind(action_id)) {
-  case ATK_CITY:
-    if (target_city == NULL) {
-      explnat->kind = ANEK_MISSING_TARGET;
-    }
-    break;
-  case ATK_UNIT:
-    if (target_unit == NULL) {
-      explnat->kind = ANEK_MISSING_TARGET;
-    }
-    break;
-  case ATK_UNITS:
-  case ATK_TILE:
-    if (target_tile == NULL) {
-      explnat->kind = ANEK_MISSING_TARGET;
+  if (action_id != ACTION_ANY) {
+    /* A specific action should have a suiting target. */
+    switch (action_id_get_target_kind(action_id)) {
+    case ATK_CITY:
+      if (target_city == NULL) {
+        explnat->kind = ANEK_MISSING_TARGET;
+      }
+      break;
+    case ATK_UNIT:
+      if (target_unit == NULL) {
+        explnat->kind = ANEK_MISSING_TARGET;
+      }
+      break;
+    case ATK_UNITS:
+    case ATK_TILE:
+      if (target_tile == NULL) {
+        explnat->kind = ANEK_MISSING_TARGET;
+      }
+      break;
+    case ATK_SELF:
+      /* No other target. */
+      break;
+    case ATK_COUNT:
+      fc_assert(action_id_get_target_kind(action_id) != ATK_COUNT);
+      break;
     }
-    break;
-  case ATK_SELF:
-    /* No other target. */
-    break;
-  case ATK_COUNT:
-    fc_assert(action_id_get_target_kind(action_id) != ATK_COUNT);
-    break;
   }
 
   if (explnat->kind == ANEK_MISSING_TARGET) {
